pkg/auth/service.go

```diff
@@ -1933,7 +1962,78 @@ func (a *APIAuthService) CheckHealth(ctx context.Context, logger logging.Logger,
 	return nil
 }
 
-func NewAPIAuthService(apiEndpoint, token string, secretStore crypt.SecretStore, cacheConf params.ServiceCache, logger logging.Logger) (*APIAuthService, error) {
+func (a *APIAuthService) IsExternalPrincipalsEnabled(ctx context.Context) bool {
+	return a.externalPrincipalseEnabled
+}
+func (a *APIAuthService) CreateUserExternalPrincipal(ctx context.Context, userID, principalID string) error {
+	if !a.IsExternalPrincipalsEnabled(ctx) {
+		return fmt.Errorf("not enabled: %w", ErrInvalidRequest)
+	}
+
+	resp, err := a.apiClient.CreateUserExternalPrincipalWithResponse(ctx, userID, principalID)
+
+	if err != nil {
+		return fmt.Errorf("create principal: %w", err)
+	}
+
+	return a.validateResponse(resp, http.StatusCreated)
+}
+
+func (a *APIAuthService) DeleteUserExternalPrincipal(ctx context.Context, userID, principalID string) error {
+	if !a.IsExternalPrincipalsEnabled(ctx) {
+		return fmt.Errorf("not enabled: %w", ErrInvalidRequest)
+	}
+	resp, err := a.apiClient.DeleteUserExternalPrincipalWithResponse(ctx, userID, principalID)
+	if err != nil {
+		return fmt.Errorf("delete external principal: %w", err)
+	}
+	return a.validateResponse(resp, http.StatusNoContent)
+}
+
+func (a *APIAuthService) GetUserExternalPrincipal(ctx context.Context, userID, principalID string) (*model.ExternalPrincipal, error) {
+	if !a.IsExternalPrincipalsEnabled(ctx) {
+		return nil, fmt.Errorf("not enabled: %w", ErrInvalidRequest)
+	}
+	resp, err := a.apiClient.GetUserExternalPrincipalWithResponse(ctx, userID, principalID)
+	if err != nil {
+		return nil, fmt.Errorf("get external principal: %w", err)
+	}
+	if err := a.validateResponse(resp, http.StatusOK); err != nil {
+		return nil, err
+	}
+	return &model.ExternalPrincipal{
+		ID:     resp.JSON200.Id,
+		UserID: resp.JSON200.UserId,
+	}, nil
+}
+
+func (a *APIAuthService) ListUserExternalPrincipals(ctx context.Context, userID string, params *model.PaginationParams) ([]*model.ExternalPrincipal, *model.Paginator, error) {
+	if !a.IsExternalPrincipalsEnabled(ctx) {
+		return nil, nil, fmt.Errorf("not enabled: %w", ErrInvalidRequest)
+	}
+	resp, err := a.apiClient.ListUserExternalPrincipalsWithResponse(ctx, userID, &ListUserExternalPrincipalsParams{
+		Prefix: paginationPrefix(params.Prefix),
+		After:  paginationAfter(params.After),
+		Amount: paginationAmount(params.Amount),
```

Comment:
Can `params` be nil?

Reply:
nope. 
If it will be then it's a bug and it should panic.